allow org members to add EEs to an org and superusers to change EEs in an org

rebeccahhh opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
allow org members to add EEs to an org and superusers to change EEs in an org and address #9359 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 17.0.1
```

rebeccahhh committed:
add org admins as able to control EEs even if they don't have the ee_admin role for the specific ee and prevent managed_by_tower EEs from being edited/deleted

jbradberry commented on awx/main/access.py:
```diff
@@ -1332,11 +1332,13 @@ def filtered_queryset(self):
 
     @check_superuser
     def can_change(self, obj, data):
+        if self.user.is_superuser:
```
`@check_superuser` already does this

rebeccahhh replied:
🤦🏻 you right

jbradberry commented on awx/main/access.py:
```diff
@@ -1332,11 +1332,13 @@ def filtered_queryset(self):
     @check_superuser
```
In the `filtered_queryset` method right above this point, there's a query using `'execution_environment_admin_role'` that needs to change to use `'member_role'`.  This is the query that gets used to see if a user can view something.

jbradberry commented on awx/main/access.py:
```diff
@@ -1332,11 +1332,13 @@ def filtered_queryset(self):
+            return True
         if obj and obj.organization_id is None:
             raise PermissionDenied
         if self.user not in obj.organization.execution_environment_admin_role:
```
I suspect that this needs to check that the user is not in `obj.organization.admin_role` as well.

jbradberry commented on awx/main/access.py:
```diff
@@ -1325,14 +1325,14 @@ class ExecutionEnvironmentAccess(BaseAccess):
     def can_add(self, data):
         if not data:  # So the browseable API will work
-            return Organization.accessible_objects(self.user, 'execution_environment_admin_role').exists()
+            return Organization.accessible_objects(self.user, 'member_role').exists()
```
I _think_ that this change isn't correct, but am willing to be shown wrong.

kdelee replied:
I agree this looks like any member can create one

rebeccahhh replied:
it sure does, and I tested it just to verify. I'll pull it out

jbradberry commented on awx/main/access.py:
```diff
@@ -1325,7 +1325,7 @@ class ExecutionEnvironmentAccess(BaseAccess):
     def filtered_queryset(self):
         return ExecutionEnvironment.objects.filter(
-            Q(organization__in=Organization.accessible_pk_qs(self.user, 'execution_environment_admin_role')) |
+            Q(organization__in=Organization.accessible_pk_qs(self.user, 'member_role')) |
```
Probably should be `'read_role'` now that I've looked more closely again at the `Organization` model roles.  Also, I think we need to add `'execution_environment_admin_role'` to `Organization.read_role`'s parent roles.

rebeccahhh replied:
`read_role` replaced member role and added to the orgs read_role list

jbradberry commented on awx/main/access.py:
```diff
@@ -1339,7 +1339,7 @@ def can_add(self, data):
         if obj and obj.organization_id is None:
             raise PermissionDenied
-        if self.user not in obj.organization.execution_environment_admin_role:
+        if self.user not in obj.organization.execution_environment_admin_role and self.user not in obj.organization.admin_role:
```
It turns out that this change isn't necessary:

    In [15]: pprint([(u, u in o.execution_environment_admin_role) for u in User.obje
        ...: cts.all()])
    [(<User: admin>, True),
     (<User: auditor>, False),
     (<User: org-admin>, True),
     (<User: org-ee-admin>, True),
     (<User: org-member>, False),
     (<User: ordinary>, False)]

rebeccahhh replied:
alright, removed

jbradberry commented on awx/main/access.py:
```diff
@@ -1325,7 +1325,7 @@ class ExecutionEnvironmentAccess(BaseAccess):
+            Q(organization__in=Organization.accessible_pk_qs(self.user, 'member_role')) |
             Q(organization__isnull=True)
         ).distinct()
 
```
Leaving this here, since `can_add` is outside of the commentable lines of code -- it turns out that `check_related` takes an optional `mandatory` flag that I think does what we want wrt Org Admins / EE Admins not being able to create global EEs.

rebeccahhh committed:
add execution_environment_admin_role to the an organizations read role, which access.py uses for determining access to reading an ee within an organization,

add migration file for execution_env_admin role addition to read_roles within an organization,

and set check related to mandatory

jbradberry committed:
Make the managed_by_tower field read-only for EEs (similar to how we deal with it not being settable for Credentials) and add permissions checking for Org EE Admins.

can_add: gets an explicit role to check against, `'execution_environment_admin_role'`
can_change: leverages `self.check_related()` for the case where the Org is not changing, but also adds an explicit check for the EE Admin Role when the Org is changing to an explicit different Org.

shanemcd commented on awx/api/serializers.py:
```diff
@@ -1362,6 +1362,7 @@ def validate(self, attrs):
 
 class ExecutionEnvironmentSerializer(BaseSerializer):
     show_capabilities = ['edit', 'delete']
+    managed_by_tower = serializers.ReadOnlyField()
```
👍

rebeccahhh commented:
@shanemcd @jbradberry I rebased with the new changes in the main EE branch, I think we're good to `mergit`

jbradberry commented on awx/main/access.py:
```diff
@@ -1308,6 +1308,54 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
+
+    @check_superuser
+    def can_change(self, obj, data):
+        if obj.managed_by_tower is True:
```
This should probably be changed to `if obj.managed_by_tower:`.  Let's not rely on Python inlining all instances of `True` to the same object.

rebeccahhh replied:
got it :)

jbradberry commented on awx/main/access.py:
```diff
@@ -1308,6 +1308,54 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
+        ).distinct()
+
+    @check_superuser
+    def can_add(self, data):
```
We should probably do a similar `managed_by_tower` check here as well.

jbradberry committed:
Turn off permissions check bypassing for admins when hitting the execution environment list and detail views.

jbradberry commented on awx/main/access.py:
```diff
@@ -1325,31 +1325,35 @@ class ExecutionEnvironmentAccess(BaseAccess):
-    @check_superuser
     def can_change(self, obj, data):
+        if obj.managed_by_tower:
+            raise PermissionDenied
```
We should also check, I think, whether `'managed_by_tower'` is being attempted to be set in the data dict.  Maybe.

jbradberry replied:
never mind, we set it to a read-only field :facepalm:

rebeccahhh commented:
regate
